debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java

```diff
@@ -491,24 +491,33 @@ protected HistoryRecordComparator getHistoryRecordComparator() {
         return new HistoryRecordComparator() {
             @Override
             protected boolean isPositionAtOrBefore(Document recorded, Document desired) {
-                Long recordedScn;
-                Long desiredScn;
+                Scn recordedScn;
+                Scn desiredScn;
                 if (getAdapter() == OracleConnectorConfig.ConnectorAdapter.XSTREAM) {
                     final LcrPosition recordedPosition = LcrPosition.valueOf(recorded.getString(SourceInfo.LCR_POSITION_KEY));
                     final LcrPosition desiredPosition = LcrPosition.valueOf(desired.getString(SourceInfo.LCR_POSITION_KEY));
-                    recordedScn = recordedPosition != null ? recordedPosition.getScn() : recorded.getLong(SourceInfo.SCN_KEY);
-                    desiredScn = desiredPosition != null ? desiredPosition.getScn() : desired.getLong(SourceInfo.SCN_KEY);
+                    recordedScn = recordedPosition != null ? recordedPosition.getScn() : resolveScn(recorded);
+                    desiredScn = desiredPosition != null ? desiredPosition.getScn() : resolveScn(desired);
                     return (recordedPosition != null && desiredPosition != null)
                             ? recordedPosition.compareTo(desiredPosition) < 1
                             : recordedScn.compareTo(desiredScn) < 1;
                 }
                 else {
-                    recordedScn = recorded.getLong(SourceInfo.SCN_KEY);
-                    desiredScn = desired.getLong(SourceInfo.SCN_KEY);
+                    recordedScn = resolveScn(recorded);
+                    desiredScn = resolveScn(desired);
                     return recordedScn.compareTo(desiredScn) < 1;
                 }
 
             }
+
+            private Scn resolveScn(Document document) {
+                // prioritize reading scn as string and if not found, fallback to long data types
+                if (document.getString(SourceInfo.SCN_KEY) == null) {
+                    Long scnValue = document.getLong(SourceInfo.SCN_KEY);
+                    Scn.valueOf(scnValue == null ? 0 : scnValue);
+                }
+                return Scn.valueOf(document.getString(SourceInfo.SCN_KEY));
```

Comment:
Let's avoid the duplicated `getString()` call by assigning it to a local var above.